### PR TITLE
Use native code on Darwin

### PR DESCRIPTION
**Pull request by danyisill**

### CODE REVIEW

**Comment by woodruffw** [comment]
Nice, thank you! I'll review these changes over the weekend.

File: src/plat/darwin/detect.c [hunk 1]
```diff
@@ -40,24 +45,22 @@
...
-	pclose(distro_file);
-
-	snprintf(distro_str, MAX_STRLEN, "Mac OS X %s", distro_vers_str);
-#endif
+	char *codenames[] = {"Cheetah", "Puma", "Jaguar", "Panther", "Tiger", "Leopard", "Snow Leopard", "Lion", "Mountain Lion", "Mavericks", "Yosemite", "El Capitan", "Sierra", "High Sierra", "Mojave"};
+	CFArrayRef split = CFStringCreateArrayBySeparatingStrings(NULL, CFPreferencesCopyAppValue(CFSTR("ProductVersion"), CFSTR("/System/Library/CoreServices/SystemVersion")), CFSTR("."));
+	unsigned maj = CFStringGetIntValue(CFArrayGetValueAtIndex(split, 0));
+	unsigned min = CFStringGetIntValue(CFArrayGetValueAtIndex(split, 1));
+	unsigned fix = 0;
+	if(CFArrayGetCount(split) == 3)
```
Comment (woodruffw) [inline src/plat/darwin/detect.c]: Style nit (here are throughout the changes): Put a space after the `if` token and always use braces on branches.

File: src/plat/darwin/detect.c [hunk 2]
```diff
@@ -40,24 +45,22 @@
...
+		fix = CFStringGetIntValue(CFArrayGetValueAtIndex(split, 2));
+
+	char build_ver[16];
+	CFStringGetCString(CFPreferencesCopyAppValue(CFSTR("ProductBuildVersion"), CFSTR("/System/Library/CoreServices/SystemVersion")), build_ver, 16, kCFStringEncodingUTF8);
+	
+	char *codename = "Unknown";
+	if(min < sizeof codenames / sizeof *codenames)
```
Comment (woodruffw) [inline src/plat/darwin/detect.c]: Nit: Prefer `sizeof(codenames)`, and `sizeof(*codenames)`.

File: src/plat/darwin/detect.c [hunk 3]
```diff
@@ -159,19 +162,9 @@ void detect_pkgs(void)
...
-		sysctlbyname("machdep.cpu.brand_string", str, &len, NULL, 0);
-	*/
-	cpu_file = popen("sysctl -n machdep.cpu.brand_string 2> /dev/null | "
-				"sed 's/(\\([Tt][Mm]\\))//g;s/(\\([Rr]\\))//g;s/^//g' | "
-				"tr -d '\\n' | tr -s ' '", "r");
-	fgets(cpu_str, MAX_STRLEN, cpu_file);
-	pclose(cpu_file);
-
+	size_t size;
+	sysctlbyname("machdep.cpu.brand_string", NULL, &size, NULL, 0);
```
Comment (woodruffw) [inline src/plat/darwin/detect.c]: Nit: needs a check for `size < MAX_STRLEN` here after the first `sysctlbyname`.

File: src/plat/darwin/detect.c [hunk 4]
```diff
@@ -180,13 +173,24 @@ void detect_cpu(void)
...
+		io_registry_entry_t regEntry;
+		while ((regEntry = IOIteratorNext(iterator))) {
+			CFMutableDictionaryRef serviceDictionary;
+			if (IORegistryEntryCreateCFProperties(regEntry, &serviceDictionary, kCFAllocatorDefault, kNilOptions) != kIOReturnSuccess){
+                	    IOObjectRelease(regEntry);
+                	    continue;
+                	}
+                	const void *GPUModel = CFDictionaryGetValue(serviceDictionary, CFSTR("model"));
+                	if (GPUModel && CFGetTypeID(GPUModel) == CFDataGetTypeID())
+					safe_strncpy(gpu_str, (char *)CFDataGetBytePtr((CFDataRef) GPUModel), MAX_STRLEN);
```
Comment (woodruffw) [inline src/plat/darwin/detect.c]: Nit: Inconsistent indentation here and below.

File: src/plat/darwin/detect.c [hunk 5]
```diff
@@ -221,27 +225,22 @@ void detect_disk(void)
...
+	vm_statistics64_data_t vm_stats;
+	mach_port = mach_host_self();
+	count = sizeof(vm_stats) / sizeof(natural_t);
+	if (KERN_SUCCESS == host_page_size(mach_port, &page_size) &&
+	KERN_SUCCESS == host_statistics64(mach_port, HOST_VM_INFO, (host_info64_t)&vm_stats, &count)){
+		used_mem = ((int64_t)vm_stats.active_count + (int64_t)vm_stats.wire_count) * (int64_t)page_size;
+	}
 
-	used_mem = total_mem - free_mem;
+	size_t len = sizeof total_mem;
```
Comment (woodruffw) [inline src/plat/darwin/detect.c]: Nit: `sizeof(total_mem)`

File: src/plat/darwin/detect.c [hunk 6]
```diff
@@ -317,13 +316,13 @@ void detect_shell(void)
...
-				"awk '/Resolution:/ {print $2\"x\"$4}' | tr -d '\\n'", "r");
-	fgets(res_str, MAX_STRLEN, res_file);
-	pclose(res_file);
-
+	uint32_t count = 0, chars = 0;
+	CGGetOnlineDisplayList(UINT32_MAX, NULL, &count);
+	CGDirectDisplayID displays[count];
+	CGGetOnlineDisplayList(count, displays, &count);
+	chars += snprintf(res_str, MAX_STRLEN, "%zu x %zu", CGDisplayPixelsWide(*displays), CGDisplayPixelsHigh(*displays));
+	for(int i = 1; i < count; ++i)
```
Comment (woodruffw) [inline src/plat/darwin/detect.c]: Nit: Always use braces around loops.

File: src/plat/darwin/detect.c [hunk 7]
```diff
@@ -351,12 +350,17 @@ void detect_wm(void)
...
 {
-	safe_strncpy(wm_theme_str, "Aqua", MAX_STRLEN);
-
+	char *accents[] = {"Graphite", "Red", "Orange", "Yellow", "Green", "", "Purple", "Pink", "Blue"};
+	Boolean e;
+	CFStringRef def = CFSTR(".GlobalPreferences");
+	CFIndex accent = CFPreferencesGetAppIntegerValue(CFSTR("AppleAccentColor"), def, &e);
+	CFIndex colorvar = CFPreferencesGetAppIntegerValue(CFSTR("AppleAquaColorVariant"), def, NULL);
+	CFPropertyListRef style = CFPreferencesCopyAppValue(CFSTR("AppleInterfaceStyle"), def);
+	snprintf(wm_theme_str, MAX_STRLEN, "%s %s", style?"Dark":"Light", e?accents[accent + 1]:colorvar == 6?"Graphite":"Blue");
```
Comment (woodruffw) [inline src/plat/darwin/detect.c]: Nit: Use spaces to make ternary expressions easier to read. Also, use temporary variables for them here -- no need for them to be inlined.

**Comment by danyisill** [comment]
Thanks, I think I fixed that

**Comment by woodruffw** [comment]
Thanks! I'll take another look tomorrow (and hopefully merge it then).

**Comment by danyisill** [comment]
Hey, did you forget to do it?

**Comment by woodruffw** [comment]
No, I’ve just been a little busy with other things. Sorry. I’ll be able to review it this weekend.

**Comment by woodruffw** [comment]
Thanks again! I'll merge this in a bit.

**Pull request merged**